### PR TITLE
[SPARK-35331][SQL] Support resolving missing attrs for distribute/cluster by/repartition hint

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR makes the below case work well. 

```sql
select a b from values(1) t(a) distribute by a;
```

```logtalk
== Parsed Logical Plan ==
'RepartitionByExpression ['a]
+- 'Project ['a AS b#42]
   +- 'SubqueryAlias t
      +- 'UnresolvedInlineTable [a], [List(1)]

== Analyzed Logical Plan ==
org.apache.spark.sql.AnalysisException: cannot resolve 'a' given input columns: [b]; line 1 pos 62;
'RepartitionByExpression ['a]
+- Project [a#48 AS b#42]
   +- SubqueryAlias t
      +- LocalRelation [a#48]
```
### Why are the changes needed?

bugfix

### Does this PR introduce _any_ user-facing change?

yes, the original attributes can be used in `distribute by` / `cluster by` and hints like `/*+ REPARTITION(3, c) */`

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon thanks for reviewing

**Comment by cloud-fan** [comment]
is this a regression? It seems to me that this should fail, as it's similar to `sql("select a b from values(1) t(a)").repartitionBy("a")`

**Comment by yaooqinn** [comment]
> is this a regression?

this is not a regression, but hive supports it.

> It seems to me that this should fail, as it's similar to `sql("select a b from values(1) t(a)").repartitionBy("a")`

For resolving attributes,  `distribute by` and `cluster by`  clauses should behave the same as  `sort by` and `group by`?

**Comment by cloud-fan** [comment]
> distribute by and cluster by clauses should behave the same as sort by and group by?

Then the plan should be
```
Project
  Repartition
    TableScan
``` 

Is this the case today?

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Hi, @yaooqinn .
As you mentioned, this is not a regression and didn't work before. I believe this should be merged as a new improvement to Apache Spark 3.2.0.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4100,6 +4100,17 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       }
     }
   }
+
+  test("SPARK-35331: Fix resolving original expression in RepartitionByExpression after aliased") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4104]: If you don't mind, could you add a test case for HINT explicitly?

**Comment by dongjoon-hyun** [comment]
Thank you for adding a test, @yaooqinn .

**Comment by yaooqinn** [comment]
> Thank you for adding a test, @yaooqinn .

It's my pleasure

**Comment by yaooqinn** [comment]
> Hi, @yaooqinn .
> As you mentioned, this is not a regression and didn't work before. I believe this should be merged as a new improvement to Apache Spark 3.2.0.

OK~

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @yaooqinn and @cloud-fan .

**Pull request closed without merging**